### ISSUE TITLE
`select array_concat([])` panicked

### ISSUE THREAD

**Issue by jonahgao** [apache/datafusion#10200]
### Describe the bug

thread 'main' panicked at datafusion/datafusion/functions-array/src/concat.rs:275:32:
index out of bounds: the len is 0 but the index is 0

### To Reproduce

Run any of the following queries in CLI.
```sh
DataFusion CLI v37.1.0
> select array_concat([]);
```
```sh
DataFusion CLI v37.1.0
>  select array_concat(make_array());
```

**Comment by Lordworms** [comment]
I can fix this one if no one is working on this

**Comment by jayzhan211** [comment]
@Lordworms Not sure what is your idea to solve this problem. I expect this to be solved with the correct signature.
I had done it at https://github.com/apache/datafusion/pull/8594. If you come out with another better idea, it is also nice.

**Comment by Lordworms** [comment]
> ```shell
> select array_concat(make_array());
> ```

Hi @jayzhan211 I was going to implement the same behavior like DuckDB
<img width="807" alt="image" src="https://github.com/apache/datafusion/assets/48054792/13d55812-e797-4199-8a94-fea01effafdb">
I could directly check the returnType here do match the behavior of DuckDB
<img width="881" alt="image" src="https://github.com/apache/datafusion/assets/48054792/6fc09205-ba3e-4961-8504-888aa65893e4">

and make datafusion support query like
<img width="482" alt="image" src="https://github.com/apache/datafusion/assets/48054792/ef59ce43-7d1e-46f0-846b-286bbd94baec">

**Comment by jayzhan211** [comment]
@Lordworms 
I think arguments validity check are more like `signature` instead of one that need to be computed by the user in `return_type`.

**Comment by Lordworms** [comment]
> @Lordworms I think arguments validity check are more like `signature` instead of one that need to be computed by the 
Yes, I understand, but if we need to validate argument in signature a suitable way to do it is more likely what you did in https://github.com/apache/datafusion/pull/8594 
user in `return_type`.
I'll try to see if I can come up with a better way, thanks for your advice!

**Comment by jayzhan211** [comment]
After #10439, I believe `array_concat` should use user-defined signature given it's speciality, therefore we can check the empty array in `coerce_types` for `array_concat` and return exec_error instead of panicking

cc @Lordworms

**Comment by jayzhan211** [comment]
I have a PR that updates the array_concat signature, maybe I can resolve the panic issue together.

**Comment by jayzhan211** [comment]
Actually, I'm thinking about whether we should change the behavior of array_concat similar to postgres and duckdb.
It is one of the earliest array functions that we don't follow what postgres and duckdb's behavior 🤔

**Comment by jonahgao** [comment]
Fixed by #10790 10790
